component/core/src/main/java/org/exoplatform/social/core/upgrade/UserPasswordHashMigration.java

```diff
@@ -0,0 +1,157 @@
+    this.picketLinkIDMService = picketLinkIDMService;
+    this.secureRandomService = secureRandomService;
+  }
+
+  @Override
+  public void processUpgrade(String s, String s1) {
+    LOG.info("Start upgrade of users passwords hashing algorithm");
+    long startupTime = System.currentTimeMillis();
+
+    PortalContainer container = PortalContainer.getInstance();
+    AtomicInteger updatedPasswords = new AtomicInteger();
+    EntityManager entityManager = this.entityManagerService.getEntityManager();
+    int totalSize;
+    try {
+      String sqlString = "SELECT jbid_io.NAME, jbid_io_creden.TEXT  FROM"
+          + " (jbid_io_creden INNER JOIN jbid_io ON jbid_io_creden.IDENTITY_OBJECT_ID = jbid_io.ID)"
+          + " INNER JOIN (SELECT jbid_io_attr.IDENTITY_OBJECT_ID FROM jbid_io_attr"
+          + " WHERE jbid_io_attr.IDENTITY_OBJECT_ID NOT IN (SELECT jbid_io_attr.IDENTITY_OBJECT_ID FROM jbid_io_attr WHERE NAME = 'passwordSalt128')"
+          + " GROUP BY jbid_io_attr.IDENTITY_OBJECT_ID) jia ON jbid_io_creden.IDENTITY_OBJECT_ID  = jia.IDENTITY_OBJECT_ID;";
+
+      RequestLifeCycle.begin(container);
+      Query nativeQuery = entityManager.createNativeQuery(sqlString);
+      List<Object[]> result = nativeQuery.getResultList();
+      AtomicLong chunckStartTime = new AtomicLong(System.currentTimeMillis());
+      updatedPasswords.set(0);
+      List<String> userWithErrors = new ArrayList<>();
+      totalSize = result.size();
+      result.forEach(item -> {
+        String userName = (String) item[0];
+        String passwordHash = (String) item[1];
+
+        try {
+          String saltString = Hex.encodeHexString(generateRandomSalt());
+          User user = picketLinkIDMService.getIdentitySession().getPersistenceManager().findUser(userName);
+          if (user != null && picketLinkIDMService.getExtendedAttributeManager().getAttribute(userName, PASSWORD_SALT_USER_ATTRIBUTE) == null) {
+            picketLinkIDMService.getExtendedAttributeManager().addAttribute(userName, PASSWORD_SALT_USER_ATTRIBUTE, saltString);
+            picketLinkIDMService.getExtendedAttributeManager().updatePassword(user, passwordHash);
+          }
+          int current = updatedPasswords.incrementAndGet();
+          if (current % CHUNCK == 0) {
+            if (picketLinkIDMService.getIdentitySession().getTransaction() != null) {
+              picketLinkIDMService.getIdentitySession().getTransaction().commit();
+            }
+            LOG.info("{}/{} passwords have been updated ({} ms for the chunck)",
+                     current,
+                     totalSize,
+                     System.currentTimeMillis() - chunckStartTime.get());
+            chunckStartTime.set(System.currentTimeMillis());
+          }
+        } catch (Exception e) {
+          userWithErrors.add(userName);
+          LOG.error("Error while creating attribute salt and updating password hash for user : {}", userName, e);
+        }
```

Comment:
Move this into a dedicated method with `@ExoTransactional` to ensure committing every change apart (`jcabi-plugin` is already in `pom.xml`, so the annotation should work).

Reply:
done